contracts/mainnet/connectors/notional/main.sol

```diff
@@ -0,0 +1,671 @@
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        tokensToRedeem = getNTokenRedeemAmount(
+            currencyId,
+            tokensToRedeem,
+            getId
+        );
+
+        BalanceActionWithTrades[] memory action = new BalanceActionWithTrades[](
+            1
+        );
+        action[0].actionType = DepositActionType.RedeemNToken;
+        action[0].currencyId = currencyId;
+        action[0].depositActionAmount = tokensToRedeem;
+        // Withdraw amount, withdraw cash balance and redeemToUnderlying are all 0 or false
+
+        bytes32[] memory trades = new bytes32[](1);
+        trades[0] = encodeLendTrade(marketIndex, fCashAmount, minLendRate);
+        action[0].trades = trades;
+
+        notional.batchBalanceAndTradeAction(address(this), action);
+
+        _eventName = "LogRedeemNTokenAndDeleverage(address,uint16,uint96,uint8,uint88)";
+        _eventParam = abi.encode(
+            address(this),
+            currencyId,
+            tokensToRedeem,
+            marketIndex,
+            fCashAmount
+        );
+    }
+
+    /**
+     * @notice Deposit asset or underlying tokens and mint nTokens in a single transaction
+     * @dev This spell allows users to deposit and mint nTokens (providing liquidity)
+     * @param currencyId notional defined currency id to deposit
+     * @param depositAmount amount of tokens to deposit
+     * @param useUnderlying if true, will accept a deposit in the underlying currency (i.e DAI), if false
+     * will use the asset currency (i.e. cDAI)
+     * @param getId id of depositAmount
+     * @param setId id to set the value of nToken balance change
+     */
+    function depositAndMintNToken(
+        uint16 currencyId,
+        uint256 depositAmount,
+        bool useUnderlying,
+        uint256 getId,
+        uint256 setId
+    )
+        external
+        payable
+        returns (string memory _eventName, bytes memory _eventParam)
+    {
+        depositAmount = getDepositAmountAndSetApproval(
+            getId,
+            currencyId,
+            useUnderlying,
+            depositAmount
+        );
+
+        BalanceAction[] memory action = new BalanceAction[](1);
+        action[0].actionType = useUnderlying
+            ? DepositActionType.DepositUnderlyingAndMintNToken
+            : DepositActionType.DepositAssetAndMintNToken;
+        action[0].currencyId = currencyId;
+        action[0].depositActionAmount = depositAmount;
+        // withdraw amount, withdraw cash and redeem to underlying are all 0 and false
+
+        uint256 nTokenBefore = getNTokenBalance(currencyId);
+        uint256 msgValue = getMsgValue(
+            currencyId,
+            useUnderlying,
+            depositAmount
+        );
+
+        notional.batchBalanceAction{value: msgValue}(address(this), action);
+
+        uint256 nTokenBalanceChange = sub(getNTokenBalance(currencyId), nTokenBefore);
+
+        if (setId != 0) {
```

Comment:
Why this? Can't we pass setId as 0?

Reply:
This allows the caller to save some gas by setting setId to zero if the get/set mechanism is not used.

Reply:
@thrilok209 Is this valid?